### DISCUSSION TITLE
Spring MVC의 실행 흐름에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#152] [Spring MVC]
![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/ec750888-8e1c-40e0-9d2d-0c838c0ca4df)

**Comment by J-I-H-O** [comment]
**<View를 응답하는 경우>**

![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/162905b5-b2d0-4472-aa8f-72b1f464894c)

1. 클라이언트로부터 HTTP 요청이 들어옵니다. 이때 DispatcherServlet이 Front Controller의 역할을 수행합니다.
2. HandlerMapping을 통해 URL에 매핑된 핸들러(컨트롤러)를 조회합니다.
3. DispatcherServlet은 찾은 핸들러를 실행하기 위해 HandlerAdapter를 사용합니다. 해당 핸들러를 실행할 수 있는 HandlerAdapter를 조회합니다.
4. HandlerAdapter가 실제로 요청을 처리하는 컨트롤러 메서드를 호출합니다.
컨트롤러는 서비스 계층과 데이터 접근 계층을 호출해 비즈니스 로직을 수행합니다.
5. 컨트롤러는 결과 데이터를 Model 객체에 담고, View 이름을 반환합니다.
6. ViewResolver를 찾아 실행합니다. 이때 ViewResolver는 View 이름을 기반으로 적절한 뷰 템플릿을 찾습니다.
7. ViewResolver가 찾은 뷰 템플릿을 사용해 최종적으로 HTML과 같은 응답을 생성합니다.
8. 최종적으로 생성된 뷰는 클라이언트에게 응답으로 반환됩니다.

---

**<Message Converter의 동작>**

![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/36d54219-4302-4b79-9fc4-96129d5dec85)

JSON과 문자열 등을 응답하는 경우는 View를 응답하는 경우와 유사한 흐름으로 동작합니다. 
단, ViewResolver 대신 HttpMessageConverter 가 동작합니다.
MessageConverter는 HTTP 요청과 응답 두 경우 모두 사용될 수 있습니다. 
이때 클라이언트의 HTTP Accept 헤더와 반환 타입 정보, Content-Type 등을 조합하여 타입에 맞는 HttpMessageConverter가 선택됩니다.

1. 클라이언트로부터 HTTP 요청이 들어옵니다. 이때 DispatcherServlet이 Front Controller의 역할을 수행합니다.
2. HandlerMapping을 통해 URL에 매핑된 핸들러(컨트롤러)를 조회합니다.
3. `@RequestMapping` 을 처리하는 핸들러 어댑터인 RequestMappingHandlerAdapter가 ArgumentResolver를 호출해 핸들러가 필요로 하는 파라미터의 값을 생성하고, 컨트롤러 메서드를 호출하면서 값을 넘겨줍니다.
이때 ArgumentResolver가 HttpMessageConverter를 사용해 필요한 객체를 생성합니다.
4. 컨트롤러는 전달받은 파라미터를 사용하여 서비스 계층과 데이터 접근 계층을 호출해 비즈니스 로직을 수행합니다.
5. `@ResponseBody` 와 `HttpEntity` 를 처리하는 ReturnValueHandler가 HTTPMessageConverter를 호출해 응답 결과를 만들어냅니다.

참고: '그 강의'

**Comment by le2sky** [comment]
스프링 MVC는 다음과 같은 흐름으로 요청을 처리합니다.

1. DispatcherServlet으로 클라이언트의 요청(HttpServletRequest)이 들어옵니다. (WAS에서 생성)
2. 웹 요청을 LocaleResolver, ThemeResolver, MultipartResolver 인터페이스의 구현체가 분석합니다.
3. 웹 요청을 처리할 핸들러를 HandlerMapping을 통해서 탐색합니다.
4. 찾은 핸들러를 실행할 수 있는 HandlerAdapter를 탐색합니다.
5. HandlerAdapter를 사용해 핸들러의 메서드를 실행하고, 이때 반환되는 값은 ModelAndView입니다.
6. View 이름을 ViewResolver에게 전달하여 해당하는 View 객체를 반환합니다. (view 이름으로 viewResolver 순서대로 호출해서 resolver 찾음)
7. DispatcherServlet은 View에게 Model을 전달하고, 화면 표시를 요청합니다. (view.render 호출)
8. 최종적으로 DispatcherServlet은 View 결과(HttpServletResponse)를 클라이언트에게 반환합니다.

RestController를 사용하는 경우에는 6,7번 과정이 생략됩니다.
ViewResolver를 사용하지 않고, 반환값에 맞는 MessageConverter를 찾아서 응답 본문 생성하여 
클라이언트에게 내려줍니다. (입력 = ArgumentResolver가 MessageConverter를 사용, 출력 = ReturnValueHandler가 MessageConverter를 사용, 위 땡이 그림 참고)

https://terasolunaorg.github.io/guideline/5.0.1.RELEASE/en/Overview/SpringMVCOverview.html#overview-of-spring-mvc-processing-sequence
https://tecoble.techcourse.co.kr/post/2021-06-25-dispatcherservlet-part-1/
https://tecoble.techcourse.co.kr/post/2021-07-15-dispatcherservlet-part-2/
그 강의
